crates/parser/src/grammar/type_args.rs

```diff
@@ -65,17 +65,48 @@ fn generic_arg(p: &mut Parser) {
             m.complete(p, LIFETIME_ARG);
         }
         // test associated_type_bounds
-        // fn print_all<T: Iterator<Item: Display>>(printables: T) {}
-        IDENT if p.nth(1) == T![:] && p.nth(2) != T![:] => {
+        // fn print_all<T: Iterator<Item, Item::Item, Item: Display, Item<'a> = Item>>(printables: T) {}
+        IDENT if [T![<], T![=], T![:]].contains(&p.nth(1)) => {
+            let path_ty = p.start();
+            let path = p.start();
+            let path_seg = p.start();
             name_ref(p);
-            type_params::bounds(p);
-            m.complete(p, ASSOC_TYPE_ARG);
-        }
-        IDENT if p.nth(1) == T![=] => {
-            name_ref(p);
-            p.bump_any();
-            types::type_(p);
-            m.complete(p, ASSOC_TYPE_ARG);
+            if p.current() == T![<] {
+                opt_generic_arg_list(p, false);
+            }
+            if p.current() == T![=] {
+                // NameRef<...> =
+                p.bump_any();
+                types::type_(p);
+
+                path_seg.abandon(p);
+                path.abandon(p);
+                path_ty.abandon(p);
+                m.complete(p, ASSOC_TYPE_ARG);
+            } else if p.current() == T![:] {
```

Comment:
We've got a chain of `p.current()` comparisons against constants here. A `match IDENT` against the macro-ized constants might be more fitting.